### ISSUE TITLE
fish should autocomplete dir-related commands with existing folder names, not with regular file names, named pipes names…

### ISSUE THREAD

**Issue by PenegalECI** [fish-shell/fish-shell#4034]
Hello, there.

```
root@hermes /etc# echo $version
2.2.0
```

I noticed that, when I try to autocomplete after a `mkdir`, fish suggestions include non-directory files, while they don't make sense in this context:
```
root@hermes /e/postfix# cd /tmp/fish.root/
root@hermes /t/fish.root# ls -lah
total 8,0K
drwx------ 2 root root 4,0K mai   16 14:09 ./
drwxrwxrwt 9 root root 4,0K mai   16 14:35 ../
prw------- 1 root root    0 mai   16 14:09 fishd.fa163e811365.notifier|
root@hermes /t/fish.root# mkdir f
```

Here, fish automatically suggests `fishd.fa163e811365.notifier`, but, even if creating a folder with the same name than a named pipe may be possible, it would be highly discouraged, so suggesting it is merely useless. In this context, fish should, IMHO, suggest only matching folder names, not other file type names. `rmdir` is also affected, and it is likely that other folder-related commands could benefit from this enhancement. 

Regards.

**Comment by faho** [comment]
`rmdir` is a simple case. It should only complete existing directories, and it is already supposed to do, but runs into #112, which should hopefully be fixed after 2.6.0.

`mkdir` isn't simple. The problem there is that completing a sensible argument for it is basically impossible - the only arguments that make sense are paths that _don't exist yet_. That can't be sensibly done (either you'd offer every one of the milliions of possible paths), so we should fall back to the best we can do.

The best we can do actually seems to be existing paths, so you can construct something based on one. And for that it seems sensible to also include files, since you can make a dirname based on a filename - e.g. there is the /etc/X11/xorg.conf file and the /etc/X11/xorg.conf.d directory.

And there's no difference in mkdir's handling of existing directories vs existing files - you can't `mkdir /tmp` either!

So, in conclusion: The rmdir part of this is included in #112, I am rejecting the mkdir part.

I hope that makes sense.

**Comment by PenegalECI** [comment]
Well, suggesting folders in `mkdir` does make sense to me, as this command is also used to create folders inside folders, not only in the working directory: if the current directory contains one named `folder1`, I can `mkdir folder1/folder2`; this way, I don't have to `cd folder1; mkdir folder2`. This is not an uncommon way to mkdir from the current directory instead of going to the parent directory of the one to be created before creating it; this way, directory suggestions are useful, as only the last element of the path does not exist yet.

**Comment by krader1961** [comment]
> Well, suggesting folders in mkdir does make sense to me...

Note that bash (at least without any customizations) behaves like fish and zsh behaves as @PenegalECI prefers. I think a case can be made for both behaviors. The question is which is most useful and least surprising. Personally, I prefer the current behavior for the same reason outlined by @faho. But I wouldn't veto changing it to behave like zsh if a convincing argument was put forth.
